plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -127,3 +154,235 @@ counted out already so we should not immediately encounter those in `VBody`.
+                fullyAppliedAnns = fmap annotation filteredFullyApplied
+              -- add the function to `CalledVarEnv`
+              -- add the type abstraction to `CalledVarEnv`
+              void $ modify' $ extendCalled n (MkCalledVarInfo rhs varLamOrder fullyAppliedAnns)
+              -- we don't remove the binding because we decide *at the call site* whether we want to
+              -- inline, and it may be called more than once
+              pure $ TermBind a s v rhs
+    -- For anything else, just process all the subterms
+    b -> forMOf bindingSubterms b processTerm
+
+-- | Counts the type and term lambdas in the RHS of a binding and returns an ordered list
+countLam ::
+    Term tyname name uni fun a -- ^ the RHS of the let binding
+    -> TermOrTypeOrder
+countLam = countLamInner []
+    where
+      countLamInner ::
+        TermOrTypeOrder
+        -> Term tyname name uni fun a -- ^ The rhs term that is being counted.
+        -> TermOrTypeOrder
+      countLamInner lamStack (LamAbs _a _n _ty body) =
+        -- If the term is a term lambda abstraction, add it to the list, and
+        -- keep on examining the body.
+        countLamInner (lamStack <> [MkTerm]) body
+      countLamInner lamStack (TyAbs _a _n _kind body) =
+        -- If the term is a type lambda abstraction, add it to the list, and
+        -- keep on examining the body.
+        countLamInner (lamStack <> [MkType]) body
+      countLamInner lamStack _ =
+        -- Whenever we encounter a body that is not a lambda abstraction, we are done counting
+        lamStack
+
+-- | A record for each call sites of a variable.
+data ApplicationOrder ann =
+  MkApplicationOrder {
+    annotation         :: ann -- ^ The annotation of the call site.
+    , applicationOrder :: TermOrTypeOrder -- ^ The sequence of term of type applications applied to
+    -- this call site.
+  }
+
+-- | A mapping of a variable's unique name to a list of `ApplicationOrder` for each of its call
+-- sites.
+type ApplicationMap ann name = Map.Map name [ApplicationOrder ann]
+
+-- | Counts the number of type and term applications in the let body, including all its subterms,
+-- and returns an `ApplicationMap`.  See note [Identifying fully applied call sites].
+countApp ::
+    Ord name => -- needed for `Map`
+    Term tyname name uni fun ann -- ^ The `body` of the `Let` term.
+    -> ApplicationMap ann name
+countApp = countLocal [] Map.empty
+    where
+      countLocal ::
+        Ord name =>
+        TermOrTypeOrder -- ^ The application order.
+        -> ApplicationMap ann name -- ^ The stack of called variables.
+        -> Term tyname name uni fun ann -- ^ The rhs term that is being counted.
+        -> ApplicationMap ann name
+      countLocal appStack calledStack = go
+        where
+          go (Apply _ _ fnBody) =
+            -- If the term is a term application, add it to the application stack, and
+            -- keep on examining the body.
```

Comment:
"body" is not a thing for an `Apply` node. An `Apply` node has a function and an argument, and doesn't have a "body".